codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -357,46 +395,43 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx, ChannelPro
                     new Http2Flags().endOfStream(endStream).priorityPresent(hasPriority).paddingPresent(padding > 0);
 
             // Read the first fragment (possibly everything).
-            int nonFragmentBytes =
-                    padding + flags.getNumPriorityBytes() + flags.getPaddingPresenceFieldLength();
+            int nonFragmentBytes = padding + flags.getNumPriorityBytes() + flags.getPaddingPresenceFieldLength();
             int maxFragmentLength = maxFrameSize - nonFragmentBytes;
             ByteBuf fragment =
-                    headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength));
+                    headerBlock.readSlice(Math.min(headerBlock.readableBytes(), maxFragmentLength)).retain();
 
             // Set the end of headers flag for the first frame.
-            flags.endOfHeaders(headerBlock.readableBytes() == 0);
+            flags.endOfHeaders(!headerBlock.isReadable());
 
             int payloadLength = fragment.readableBytes() + nonFragmentBytes;
-            ByteBuf firstFrame = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
-            writeFrameHeader(firstFrame, payloadLength, HEADERS, flags,
-                    streamId);
-
-            // Write the padding length.
-            writePaddingLength(padding, firstFrame);
+            ByteBuf buf = ctx.alloc().buffer(HEADERS_FRAME_HEADER_LENGTH);
+            writeFrameHeaderInternal(buf, payloadLength, HEADERS, flags, streamId);
+            writePaddingLength(buf, padding);
 
-            // Write the priority.
             if (hasPriority) {
                 long word1 = exclusive ? 0x80000000L | streamDependency : streamDependency;
-                writeUnsignedInt(word1, firstFrame);
+                writeUnsignedInt(word1, buf);
 
                 // Adjust the weight so that it fits into a single byte on the wire.
-                firstFrame.writeByte(weight - 1);
+                buf.writeByte(weight - 1);
             }
+            ctx.write(buf, promiseAggregator.newPromise());
 
             // Write the first fragment.
-            firstFrame.writeBytes(fragment);
+            ctx.write(fragment, promiseAggregator.newPromise());
 
-            // Write out the padding, if any.
-            firstFrame.writeZero(padding);
+            if (padding > 0) { // Write out the padding, if any.
+                buf = ctx.alloc().buffer(padding).writeZero(padding);
```

Comment:
Yes this sounds like a good idea. Padding must be <= than 255 bytes so it seems reasonable.